bftengine/src/bftengine/KeyExchangeManager.cpp

```diff
@@ -229,9 +232,21 @@ void KeyExchangeManager::sendKeyExchange(const SeqNum& sn) {
     LOG_INFO(KEY_EX_LOG, "ignore request - already exchanged keys for seqnum: " << private_keys_.lastGeneratedSeqnum());
     return;
   }
+  if (sn > 0 && candidate_private_keys_.generated.sn > sn) {
+    LOG_INFO(KEY_EX_LOG,
+             "ignore request - we already have a candidate for advanced sequence number: " << KVLOG(
+                 sn, candidate_private_keys_.generated.sn));
+    return;
+  }
   KeyExchangeMsg msg;
-  if (sn == candidate_private_keys_.generated.sn && !candidate_private_keys_.generated.cid.empty()) {
-    LOG_INFO(KEY_EX_LOG, "we already have a candidate for this sequence number, trying to send it again");
+  // Notice, that a replica always sends candidates in acsending oreder, this means that in every prePrepare message,
+  // even if it consist two KEY-EXCHANGE messages from the same replica, the first one will always be possessed before
+  // the later one
+  if (sn > 0 && sn - candidate_private_keys_.generated.sn / checkpointWindowSize < 2 &&
```

Comment:
`if (sn > 0 && (sn - candidate_private_keys_.generated.sn) / checkpointWindowSize < 2 &&
      !candidate_private_keys_.generated.cid.empty()) {`

Did you mean this?

Reply:
good catch!
Thanks